### ISSUE TITLE
Alarm report issue

### ISSUE THREAD

**Issue by KihongHeo** [facebook/infer#514]
Hi Infer team,

I am making a new checker using the abstract interpretation module of Infer.
When reporting alarms, my checker sometimes ignores some of them.

Here is a brief description of the error reporting part in my checker code. 
```ocaml
let checker { Callbacks.get_proc_desc; proc_name; proc_desc; tenv; } =
  let post = Interprocedural.checker callback get_proc_desc in
  match post with
  | Some post -> 
      (* iteration on found errors *)
      Error.iter (fun e -> 
        Checkers.ST.report_error tenv
          e.proc_name             (* here, I don't use the parameter *)
          e.proc_desc             (* here, I don't use the parameter *)
          "MY CHECKER"
          e.location
          err_description
      ) errors
  | _ -> ()
```
Basically I followed the tutorial and I  think the only differences 
are the commented parts. ("proc_name" and "proc_desc")

In my experience, the alarm reporting works well 
when Infer takes a single file. But it ignores some of them when 
analyzing multiple files. 

Could you give me an advice?

Thanks,
Kihong

**Comment by akotulski** [comment]
> But it ignores some of them when
analyzing multiple files.

How are you analyzing multiple files? What are the infer commands that you run?

**Comment by KihongHeo** [comment]
I run Infer as follows:
```
./infer -a checkers -- gcc -c test1.c test2.c
```

The problem also happens
```
./infer -a checkers -- make
```

**Comment by akotulski** [comment]
That looks correct. I'll let someone more knowledgeable about writing checkers look at the question.

**Comment by jvillard** [comment]
It's hard (for me) to tell without looking at the full code and examples. For debugging, you can use `InferPrint` to print the specs computed by your checker, for instance:

```
./infer -a checkers -- make
InferPrint
```

That may give you an idea of what could be wrong.

**Comment by cristianoc** [comment]
Can you try disabling all the other checkers in `checkers/registerCheckers.ml`, to make sure there's no interference.

**Comment by KihongHeo** [comment]
Thanks for your help.

I disabled all the other checkers but the results are the same.
And the output of ```InferPrint``` is as follows:
```
Procedure: foo
int foo(char* p, int idx)
Timestamp: 3
Status: INACTIVE
Phase: FOOTPRINT
Dependency_map:
ERRORS:
WARNINGS:
TIME:0.000000 s FAILURE:NONE SYMOPS:0 CALLS:0,0

Procedure: main
int main()
Timestamp: 2
Status: INACTIVE
Phase: FOOTPRINT
Dependency_map:
ERRORS: MY CHECKER
WARNINGS:
TIME:0.000000 s FAILURE:NONE SYMOPS:0 CALLS:0,2
```

```main``` and ```foo``` are in different files,  ```foo``` has an error
and  ```main``` calls ```foo```.
It is an inter-procedural analysis and propagates safety conditions in callees to callers.
Finally it reports alarms according to the conditions at the end of ```main```.

**Comment by jvillard** [comment]
My bad, you also need to teach infer how to print your new payload for `InferPrint` to be useful. Add your payload to `Specs.pp_payload` here, like it's done for quandary, siof, threadsafety, ...: https://github.com/facebook/infer/blob/827d7c3d9ca49aedeb069131947fa574dcb2f51c/infer/src/backend/specs.ml#L448

**Comment by KihongHeo** [comment]
I see. But ```InferPrint```'s result is correct. 
I will look into the problem more deeply and give you other helpful observations.

**Comment by sblackshear** [comment]
Like @jvillard, I think it will be hard to tell without seeing more code, but here's a few ideas off the top of my head:

- Try running infer with `--no-filtering`
- Report on the `proc_name`/ `proc_desc`/`location` of the current procedure rather using `e.proc_desc` and so on. I think all of the checkers we have so far log errors that occur in a function `foo` called from `main` in `main` rather than in `foo` (we have a mechanism for displaying interprocedural traces that will show the path from `main` to `foo`). In principle, reporting the error from anywhere should work, but I haven't tried it.

@cristianoc's advice about disabling the the other checkers is also very important. I forgot to mention this in the documentation, but there's a limitation where we can only register one interprocedural checker at a time (otherwise, they clobber each other's results).

**Comment by KihongHeo** [comment]
Thanks @sblackshear . 

- Running with ```--no-filtering``` does not change anything.
- I don't get the second point clearly. My checker collects ```foo```'s safety conditions and merges them into ```main```'s safety conditions. Each safety condition (```e``` in the above code) has ```proc_name``` and ```proc_desc```. Then, it prints out  all the collected errors using ```Checkers.ST.report_error``` when the parameter ```proc_name``` is for ```main```. Can you give me a better way?

**Comment by sblackshear** [comment]
Ah, so the parameter `e.proc_name` is `main` in your example (as it should be, I think). Does the parameter `e.proc_desc` also correspond to the procdesc of `main`?

**Comment by KihongHeo** [comment]
No. I want to keep the original location of errors, so each ```e``` has its original information.

For example,
```c
void foo(int x){
  /* error1 here */
}

int main(){
  foo(x);
  /* error2 here */
}
```
I collect all the errors after analyzing ```main```, i.e., ```errors``` = {error1, error2}.
So ```error1.proc_name``` = ```foo```, ```error2.proc_name``` = ```main```, and so on. 
When I just print them out using ```prerr_endline```,  the output is correct,
but ```Checkers.ST.report_error``` misses  ```error1``` in ```foo```.

**Comment by sblackshear** [comment]
This should work, but (as mentioned above) I am not sure that it actually will. All of the other checkers we have would log error1 in `foo` while analyzing `foo` instead of deferring the logging of all errors until the analysis of `main` has finished.

I would try logging errors as soon as they are detected (if you can) and see if that fixes the problem. Alternatively, I'm happy to try debugging the issue in my end if you send a pull request or email me the code for the checker.

**Comment by KihongHeo** [comment]
Thanks a lot. I will try to look into the problem more deeply. 

By the way, the analysis on multiple files run in parallel? 
I use a reference type value (counter) to make unique symbols when analyzing each procedure.
I don't fully understand the whole Infer system so I am suspicious about the uniqueness.

**Comment by jberdine** [comment]
Yes, by default the analysis runs in parallel, unless it fails to determine the number of processors your machine has.  You can specify explicitly with `-j N`.

**Comment by KihongHeo** [comment]
Is it guaranteed that one procedure is analyzed only once?

**Comment by jberdine** [comment]
> Is it guaranteed that one procedure is analyzed only once?

No.  If procedures are recursive, they may be analyzed a few times.  Each procedure should have only a single spec written to disk, however.

**Comment by skcho** [comment]
We read more Infer code and our issue became more clear. I remind you about our situation.

1. A callee(`foo`) and a caller(`main`) are in different files.
2. A bug is in the callee.
3. When analyzing the callee, we are not sure that the code is buggy. Thus, after analyzing the callee, we do not report the bug yet.
4. After analyzing the caller, we are sure there is a bug in the callee. So we report it then.

The bug report code is as follows. Since we use the `caller_name` as the seccond argument, the bug info is added to caller(`main`)'s summary.
```ocaml
Checkers.ST.report_error
    tenv
    caller_name    (* main *)
    callee_desc    (* foo *)
    "CHECKER"
    bug_location   (* foo's file name, line number, column *)
    to_string
```

 A problem is when printing it to `report.json`.  `InferPrint.IssueJson.pp_issues_of_error_log` constructs a record, `bug`, for printing, but the location is strange; the line/column number from `foo`'s file, but the file name from `main`'s.
```java
let bug = { ...,
    line: loc.Location.line,   /* line number in foo */
    column: loc.Location.col,  /* column number in foo */
    file,                      /* main's file name */
... }
```

This causes an incorrect report. Even though, the bug is in `foo`, the file of `main` is printed.
The line number `2` is about the file of `foo`, not of `main`.
```
main.c:2: error: CHECKER
  1.   int main(){
  2. >   char* p = malloc(5);
  3.     foo(p, 10);
  4.   
```

As of now, we are using `loc.Location.file` instead of `file` when constructing the `bug` record, but not sure it is a correct way. Can you give us some advice?

**Comment by sblackshear** [comment]
I would recommend reporting an interprocedural trace starting at `main` and ending in `foo`. There's a general mechanism for creating and displaying such traces. To check out the visualization, you can run Infer (the bi-abduction analyzer) on a file with an interprocedural trace, then run `inferTraceBugs` (as described [here](http://fbinfer.com/docs/infer-workflow.html)).

To get your own nifty traces that can be visualized by `inferTraceBugs`, you'll need to create an [`Errlog.loc_trace`](https://github.com/facebook/infer/blob/master/infer/src/IR/Errlog.ml#L27). You can then pass this trace to the `ltr` parameter `Reporting.log_error` ([example](https://github.com/facebook/infer/blob/master/infer/src/quandary/TaintAnalysis.ml#L182)).

With the information you have in your current domain, I *think* you'll only be able to give the starting point and ending point of the trace (i.e., `main` and `foo` but not `callsFoo` if the call graph is `main` -> `callsFoo` -> `foo`), but that's still a great start! In the long run, it might be useful to change the domain so you can report the entire path.

**Comment by skcho** [comment]
Thank you. I will take a look at it.

**Comment by KihongHeo** [comment]
헐. 해결하셨어요?

- 허기홍 드림

2016. 12. 10. 오전 3:17 skcho <notifications@github.com> 작성:

**Comment by skcho** [comment]
@sblackshear We solved our issue by collecting the starting points and ending points of the trace as you said. It works now though we did not apply the general `TraceDomain`. Thank you for help again.

**Comment by sblackshear** [comment]
Glad to hear it! I'll close this for now then; please feel free to open another issue if you have any questions about creating a `TraceDomain`.
